insrequester/README.md

```diff
@@ -0,0 +1,73 @@
+# Requester Package
+
+The insrequester package is designed to provide a resilient way to make HTTP requests while incorporating various features like retry, circuit breaking, and timeout handling. This package aims to simplify the process of making HTTP requests by handling common scenarios like network failures, high request loads, and service timeouts, all while providing a clean and flexible API for integrating into your codebase.
+
+
+## Usage in Apps
+
+```go
+import (
+    "github.com/useinsider/insrequester"
+)
+```
+
+### Creating a Requester
+To create a new instance of the Requester, you can use the `NewRequester()` function:
+
+```go
+requester := insrequester.NewRequester()
+```
+
+### Making a Request
+The `Requester` interface provides methods for making various HTTP requests, such as GET, POST, PUT, and DELETE. Here's an example of making a GET request:
+
+```go
+requestEntity := insrequester.RequestEntity{
+    Endpoint: "https://api.example.com/resource",
+}
+
+response, err := requester.Get(requestEntity)
+if err != nil {
+    // Handle the error
+} else {
+    // Process the response
+}
+```
+
+### Adding Resilience Features
+#### Retry
+You can add retry functionality to your requests by chaining the WithRetry method to the Requester:
+
+```go
+retryConfig := insrequester.RetryConfig{
+    WaitBase: time.Millisecond * 200,
+    Times:    3,
+}
+requester.WithRetry(retryConfig)
+```
+#### Circuit Breaker
+To implement a circuit breaker pattern, use the WithCircuitbreaker method:
+
+```go
+circuitBreakerConfig := insrequester.CircuitBreakerConfig{
+    MinimumRequestToOpen:         3,
+    SuccessfulRequiredOnHalfOpen: 1,
+    WaitDurationInOpenState:      5 * time.Second,
+}
+
+requester.WithCircuitbreaker(circuitBreakerConfig)
+```
+#### Timeout
+For setting a timeout on requests, you can utilize the WithTimeout method:
+
+```go
+timeoutSeconds := 30
+requester.WithTimeout(timeoutSeconds)
```

Comment:
burada request'in kendi timeout'u olduğu, overridable olduğunu not etmek iyi olur.

Reply:
değil mi yoksa : D

Reply:
overridable olmali :D

Reply:
şöyle düşünmüştüm.
1. requester'in default timeout'u var.
2. default'u istediğin sayıya setleyebiliyorsun WithTimeout diyerek.
3. request gönderirken, sadece o request için requester'in default'unu override edebiliyorsun.

Reply:
timeout request'e degil requester'a ozel olmali bence. o yuzden her is icin ayri requester aciyoruz. kullandigimiz pakette oyle bi ozellik yok baktigim kadariyla. requester'a ne veriyosan onu kullaniyo. request'in timeout'una bakmiyo